### PR TITLE
Issue 8714: Make redirects more secure

### PR DESCRIPTION
**Pull request by annando**
Fixes https://github.com/friendica/friendica/issues/8714

We now throw errors if we detect something fishy, means:
* Fail if the visitor isn't authenticated to this site
* Fail if the contact isn't known
* Fail if the redirected URL isn't on the same host than the contact

We only use this redirect with an url with the "plink" - which is on the same host. So these checks should be okay.

### CODE REVIEW

File: mod/redir.php [hunk 1]
```diff
@@ -44,102 +47,105 @@ function redir_init(App $a) {
...
 
-	if (!empty($cid)) {
-		$fields = ['id', 'uid', 'nurl', 'url', 'addr', 'name', 'network', 'poll', 'issued-id', 'dfrn-id', 'duplex', 'pending'];
-		$contact = DBA::selectFirst('contact', $fields, ['id' => $cid, 'uid' => [0, local_user()]]);
-		if (!DBA::isResult($contact)) {
-			notice(DI::l10n()->t('Contact not found.'));
-			DI::baseUrl()->redirect();
-		}
+	if (empty($cid)) {
+		throw new \Friendica\Network\HTTPException\NotFoundException(DI::l10n()->t('Contact not found.'));
```
Comment (MrPetovan) [inline mod/redir.php:51]: This should be a `BadRequestException` instead.

File: mod/redir.php [hunk 2]
```diff
@@ -184,3 +185,20 @@ function redir_magic($a, $cid, $url)
...
+	$url_host = parse_url($url, PHP_URL_HOST);
+	$contact_url_host = parse_url($contact_url, PHP_URL_HOST);
+
+	if ($url_host == $contact_url_host) {
+		return;
+	}
+
+	Logger::error('URL check host mismatch', ['contact' => $contact_url, 'url' => $url]);
+	throw new \Friendica\Network\HTTPException\ForbiddenException(DI::l10n()->t('Access denied.'));
+}
```
Comment (MrPetovan) [inline mod/redir.php:204]: Please add an EOF EOL.

File: mod/redir.php [hunk 3]
```diff
@@ -44,102 +47,105 @@ function redir_init(App $a) {
...
-			$dest = (!empty($url) ? '&destination_url=' . $url : '');
+		$dest = (!empty($url) ? '&destination_url=' . $url : '');
 
-			System::externalRedirect($contact['poll'] . '?dfrn_id=' . $dfrn_id
-				. '&dfrn_version=' . DFRN_PROTOCOL_VERSION . '&type=profile&sec=' . $sec . $dest . $quiet);
-		}
+		System::externalRedirect($contact['poll'] . '?dfrn_id=' . $dfrn_id
+			. '&dfrn_version=' . DFRN_PROTOCOL_VERSION . '&type=profile&sec=' . $sec . $dest . $quiet);
 
+		redir_check_url($contact_url, $url);
```
Comment (MrPetovan) [inline mod/redir.php:129]: `System::externalRedirect` exits or throws an exception, nothing after it is executed.

Comment (annando) [reply mod/redir.php:129]: ?

File: mod/redir.php (hunk 1 above)
Comment (annando) [reply mod/redir.php]: Okay.

File: mod/redir.php (hunk 2 above)
Comment (annando) [reply mod/redir.php]: Done

**Comment by MrPetovan** [comment]
Awesome!

**Comment by ben-utzer** [comment]
Can we add this to the current stable version? Like a back port?

**Comment by AlfredSK** [comment]
The release is around the corner. The RC will be branched off soon, I guess.

**Comment by ben-utzer** [comment]
Yes, but that means 1 month of abuse, I guess the problem is now public and will be used by many scammers.

**Comment by MrPetovan** [comment]
You can manually apply the changes on your node until the release if you're afraid. Updating master at this point wouldn't significantly increase the security coverage. Between the develop nodes and the obsolete nodes, it would concern a minority of them. This makes me think this won't be widely used.

**Comment by AlfredSK** [comment]
I had a look at the mod/redir.php file. There were no changes from the last release until this PR. So, I copied the file from develop to a master node. Let's see how it works.

**Comment by ben-utzer** [comment]
> I had a look at the mod/redir.php file. There were no changes from the last release until this PR. So, I copied the file from develop to a master node. Let's see how it works.

Did that manually too.

**Pull request merged**